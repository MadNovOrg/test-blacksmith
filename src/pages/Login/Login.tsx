import React, { useState } from 'react'
import { useNavigate, useLocation, useSearchParams } from 'react-router-dom'
import { Link } from 'react-router-dom'
import * as EmailValidator from 'email-validator'

import { Input } from '@app/components/Input'
import { LoggedOutHeader } from '@app/components/LoggedOutHeader'

import { useAuth } from '@app/context/auth'

type SpecificErrorCodeToMessage = {
  passwordMessage: string
  emailMessage: string
}
interface Map {
  [key: string]: SpecificErrorCodeToMessage | undefined
}

const errorCodeToMesaageMapping: Map = {
  NotAuthorizedException: {
    passwordMessage: 'Incorrect username or password',
    emailMessage: 'Incorrect username or password',
  },
  UserNotFoundException: {
    passwordMessage: '',
    emailMessage: 'User does not exist',
  },
  InvalidParameterException: {
    passwordMessage: 'Please enter a password',
    emailMessage: '',
  },
  '401': {
    passwordMessage: 'New password required',
    emailMessage: '',
  },
}

type LocationState = {
  from: {
    pathname: string
  }
}

type ErrorState = {
  emailErrorMessage: string
  passwordErrorMessage: string
}

export const LoginPage = () => {
  const [searchParams] = useSearchParams()
  const navigate = useNavigate()
  const location = useLocation()
  const auth = useAuth()

  const [errorState, setErrorState] = useState<ErrorState>({
    emailErrorMessage: '',
    passwordErrorMessage: '',
  })

  const [email, setEmail] = useState(searchParams.get('email'))
  const [password, setPassword] = useState('')
  const showResetPasswordMessage =
    searchParams.get('email') &&
    searchParams.get('justResetPassword') === 'true'

  const from = (location.state as LocationState)?.from?.pathname || '/'

  const handleSubmit = async (event: React.ChangeEvent<HTMLFormElement>) => {
    event.preventDefault()

    if (!EmailValidator.validate(email as string)) {
      setErrorState(s => ({
        ...s,
        emailErrorMessage: 'Please enter a valid email address',
      }))
    } else {
      setErrorState(s => ({
        ...s,
        emailErrorMessage: '',
      }))

      const loginResult = await auth.login(email as string, password)

      // if successdfully logged in; redirect, otherwise render error message(s)
      if (loginResult.error === undefined) {
        navigate(from, { replace: true })
      } else {
        if (
          errorCodeToMesaageMapping[loginResult.error?.code || '']?.emailMessage
        ) {
          setErrorState(s => ({
            ...s,
            emailErrorMessage:
              errorCodeToMesaageMapping[loginResult.error?.code || '']
                ?.emailMessage ||
              'An error occurred. Please contact your administrator',
          }))
        }

        if (
          errorCodeToMesaageMapping[loginResult.error?.code || '']
            ?.passwordMessage
        ) {
          setErrorState(s => ({
            ...s,
            passwordErrorMessage:
              errorCodeToMesaageMapping[loginResult.error?.code || '']
                ?.passwordMessage ||
              'An error occurred. Please contact your administrator',
          }))
        }
      }
    }
  }

  return (
    <div className="flex flex-col items-center justify-center min-h-screen">
      <LoggedOutHeader />

      {showResetPasswordMessage && (
        <p className="mb-8 text-sm md:text-base">
          Please login with your new password
        </p>
      )}

      <div className="w-60 md:w-96">
        <form onSubmit={handleSubmit}>
          <section className="space-y-6 mb-16">
            <Input
              onChange={e => setEmail(e.target.value)}
              error={errorState.emailErrorMessage}
              placeholder="Please enter your email address"
              name="email"
              title="email"
              label="Email Address"
              value={email || ''}
            ></Input>

            <Input
              onChange={e => setPassword(e.target.value)}
              error={errorState.passwordErrorMessage}
              placeholder="Please enter your password"
              isPassword={true}
              name="password"
              title="password"
              label="Password"
              type="password"
            ></Input>
          </section>

          <div className="text-center">
            <button type="submit" className="btn primary w-40">
              Sign In
            </button>
          </div>
        </form>
      </div>

      <div className="mt-12">
        <Link to="/forgot-password" className="text-xs font-light underline">
          Forgotten your password?
        </Link>
      </div>
    </div>
  )
}
