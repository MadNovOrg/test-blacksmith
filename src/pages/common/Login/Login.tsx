import { yupResolver } from '@hookform/resolvers/yup'
import { LoadingButton } from '@mui/lab'
import {
  Box,
  Alert,
  FormHelperText,
  Link,
  TextField,
  Typography,
  useMediaQuery,
  useTheme,
} from '@mui/material'
import { useMemo, useState } from 'react'
import { useForm } from 'react-hook-form'
import { useTranslation } from 'react-i18next'
import {
  createSearchParams,
  useLocation,
  useNavigate,
  useSearchParams,
} from 'react-router-dom'
import * as yup from 'yup'

import { LinkBehavior } from '@app/components/LinkBehavior'
import { useAuth } from '@app/context/auth'
import { AppLayoutMinimal } from '@app/layouts/AppLayoutMinimal'

type LocationState = { from: { pathname: string; search: string } }

export type LoginInput = {
  email: string
  password: string
}

export const LoginPage = () => {
  const [searchParams] = useSearchParams()
  const navigate = useNavigate()
  const location = useLocation()
  const auth = useAuth()
  const { t } = useTranslation()

  const theme = useTheme()
  const isMobile = useMediaQuery(theme.breakpoints.down('md'))

  const [isLoading, setIsLoading] = useState(false)
  const [loginError, setLoginError] = useState<string | null>(null)

  const showResetPassMessage =
    searchParams.get('email') &&
    searchParams.get('justResetPassword') === 'true'

  const passwordResent = searchParams.get('passwordResent')
  const invitationDeclined = searchParams.get('invitationDeclined')
  const callbackUrl = searchParams.get('callbackUrl')

  const from = (location.state as LocationState)?.from || {}

  const schema = useMemo(
    () =>
      yup
        .object({
          email: yup
            .string()
            .transform(currentValue => currentValue.trim())
            .email(t('validation-errors.email-invalid'))
            .required(t('validation-errors.email-required')),
          password: yup
            .string()
            .required(t('validation-errors.password-required')),
        })
        .required(),
    [t]
  )

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<LoginInput>({
    resolver: yupResolver(schema),
    defaultValues: {
      email: searchParams.get('email') ?? '',
      password: '',
    },
  })

  const onSubmit = async (data: LoginInput) => {
    setIsLoading(true)
    setLoginError(null)

    const { user, error } = await auth.login(data.email, data.password)

    if (!error) {
      const to = `${from.pathname || '/'}${from.search || ''}`
      // https://github.com/aws-amplify/amplify-js/issues/3733
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      if (user?.challengeName === 'NEW_PASSWORD_REQUIRED') {
        return navigate({
          pathname: '/change-password',
          search: `?${createSearchParams({
            email: data.email,
          })}`,
        })
      }
      if (callbackUrl !== null) {
        window.location.replace(callbackUrl)
        return
      }
      return navigate(to, { replace: true })
    }

    setIsLoading(false)
    setLoginError(
      t(`pages.login.auth-errors.${error.code}`) ||
        t(`pages.login.auth-errors.UnknownError`)
    )
  }

  return (
    <AppLayoutMinimal>
      {showResetPassMessage ? (
        <Alert variant="outlined" severity="info" sx={{ mb: 2 }}>
          {t('pages.login.reset-pass')}
        </Alert>
      ) : null}

      {passwordResent ? (
        <Alert variant="outlined" severity="success" sx={{ mb: 2 }}>
          {t('pages.login.password-resent')}
        </Alert>
      ) : null}

      {invitationDeclined ? (
        <Alert variant="outlined" severity="success" sx={{ mb: 2 }}>
          {t('pages.login.invitation-declined')}
        </Alert>
      ) : null}

      <Typography
        variant="h3"
        sx={{ textAlign: 'center', fontWeight: 600, color: 'grey.800' }}
      >
        {t('pages.login.login-in-tt')}
      </Typography>

      <Box
        component="form"
        onSubmit={handleSubmit(onSubmit)}
        data-testid="LoginForm"
        noValidate
        autoComplete="off"
        sx={{ mt: 4 }}
      >
        <Box component="section">
          <Box mb={4}>
            <TextField
              id="email"
              variant="filled"
              label={t('pages.login.email-label')}
              placeholder={t('pages.login.email-placeholder')}
              title={t('pages.login.email-title')}
              error={!!errors.email}
              helperText={errors.email?.message}
              {...register('email')}
              autoFocus
              fullWidth
              inputProps={{ 'data-testid': 'input-email' }}
            />
          </Box>
          <Box mb={4}>
            <TextField
              id="password"
              type="password"
              variant="filled"
              label={t('pages.login.pass-label')}
              placeholder={t('pages.login.pass-placeholder')}
              title={t('pages.login.pass-title')}
              error={!!errors.password}
              helperText={errors.password?.message}
              {...register('password')}
              fullWidth
              inputProps={{ 'data-testid': 'input-password' }}
            />
          </Box>
        </Box>

        <Box display="flex" flexDirection="column" alignItems="center">
          <LoadingButton
            loading={isLoading}
            type="submit"
            variant="contained"
            color="primary"
            data-testid="login-submit"
            size="large"
            fullWidth={isMobile}
          >
            {t('pages.login.submit-label')}
          </LoadingButton>

          {loginError && (
            <FormHelperText sx={{ mt: 2 }} error data-testid="login-error">
              {loginError}
            </FormHelperText>
          )}

          <Link
            href="/registration"
            component={LinkBehavior}
            state={from ? { from } : undefined}
            data-testid="sign-up-link"
            sx={{ mt: 4, color: 'primary.main', fontWeight: '600' }}
          >
            {t('pages.login.sign-up-message')}
          </Link>
        </Box>
      </Box>

      <Box sx={{ mt: 8, textAlign: 'center' }}>
        <Link
          href="/forgot-password"
          data-testid="forgot-password-link"
          sx={{ color: 'primary.main', fontWeight: '600' }}
        >
          {t('pages.login.forgot-label')}
        </Link>
      </Box>
    </AppLayoutMinimal>
  )
}
