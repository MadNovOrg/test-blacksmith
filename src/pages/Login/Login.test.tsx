import React from 'react'
import { MemoryRouter } from 'react-router-dom'

import { LoginPage } from './Login'

import {
  render,
  providers,
  waitForCalls,
  fireEvent,
  chance,
  screen,
  waitForText,
} from '@test/index'

const mockNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}))

describe('Login', () => {
  it('matches snapshot', async () => {
    const { container } = render(
      <MemoryRouter>
        <LoginPage />
      </MemoryRouter>
    )

    expect(container).toMatchSnapshot()
  })

  it('shows error when email is empty', async () => {
    render(
      <MemoryRouter>
        <LoginPage />
      </MemoryRouter>
    )

    const email = screen.getByTestId('input-email')
    expect(email).toHaveValue('')
    expect(
      screen.queryByText('Please enter your email')
    ).not.toBeInTheDocument()

    fireEvent.click(screen.getByTestId('login-submit'))

    await waitForText('Please enter your email')
    await waitForText('Please enter your password')

    expect(providers.auth.login).not.toBeCalled()
  })

  it('shows error when email is invalid', async () => {
    render(
      <MemoryRouter>
        <LoginPage />
      </MemoryRouter>
    )

    const email = screen.getByTestId('input-email')

    fireEvent.change(email, { target: { value: 'not@valid' } })
    fireEvent.click(screen.getByTestId('login-submit'))

    await waitForText('Please enter a valid email address')

    expect(providers.auth.login).not.toBeCalled()
  })

  it('attempts login when email is valid', async () => {
    providers.auth.login.mockResolvedValue({ error: { code: 'OnPurpose' } })

    render(
      <MemoryRouter>
        <LoginPage />
      </MemoryRouter>
    )

    fireEvent.change(screen.getByTestId('input-email'), {
      target: { value: chance.email() },
    })
    fireEvent.change(screen.getByTestId('input-password'), {
      target: { value: 'secret-pass' },
    })
    fireEvent.click(screen.getByTestId('login-submit'))

    await waitForCalls(providers.auth.login, 1)
  })

  it('shows error if password is incorrect', async () => {
    providers.auth.login.mockResolvedValue({
      error: { code: 'NotAuthorizedException' },
    })

    render(
      <MemoryRouter>
        <LoginPage />
      </MemoryRouter>
    )

    const email = screen.getByTestId('input-email')
    fireEvent.change(email, { target: { value: chance.email() } })

    const pass = screen.getByTestId('input-password')

    fireEvent.change(pass, { target: { value: chance.word() } })
    fireEvent.click(screen.getByTestId('login-submit'))

    await waitForText(
      'Email address or password was incorrect, please try again'
    )
  })

  it('navigates away when login succeeds', async () => {
    providers.auth.login.mockResolvedValue({}) // no error

    render(
      <MemoryRouter>
        <LoginPage />
      </MemoryRouter>
    )

    const email = screen.getByTestId('input-email')
    fireEvent.change(email, { target: { value: chance.email() } })

    const pass = screen.getByTestId('input-password')
    fireEvent.change(pass, { target: { value: chance.word() } })

    fireEvent.click(screen.getByTestId('login-submit'))
    await waitForCalls(providers.auth.login, 1)

    expect(mockNavigate).toBeCalledWith('/', { replace: true })
  })
})
