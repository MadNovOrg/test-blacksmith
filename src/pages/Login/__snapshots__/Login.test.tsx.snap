// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-1nrbtl5"
      >
        <div
          class="MuiBox-root css-mf869q"
        >
          <div
            height="80"
            width="80"
          />
        </div>
        <div
          class="MuiBox-root css-1cb30mv"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-1ujikh7-MuiTypography-root"
          >
            Log in to Team Teach
          </h6>
          <form
            autocomplete="off"
            class="MuiBox-root css-h5fkc8"
            data-testid="LoginForm"
            novalidate=""
          >
            <section
              class="MuiBox-root css-0"
            >
              <div
                class="MuiBox-root css-1yjvs5a"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                  title="Email Address"
                >
                  <label
                    class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for="email"
                    id="email-label"
                  >
                    Email Address
                  </label>
                  <div
                    class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-ghsjzk-MuiInputBase-root-MuiInput-root"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                      data-testid="input-email"
                      id="email"
                      name="email"
                      placeholder="Please enter your email address"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                class="MuiBox-root css-1yjvs5a"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                  title="Password"
                >
                  <label
                    class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for="password"
                    id="password-label"
                  >
                    Password
                  </label>
                  <div
                    class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-ghsjzk-MuiInputBase-root-MuiInput-root"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                      data-testid="input-password"
                      id="password"
                      name="password"
                      placeholder="Please enter your password"
                      type="password"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </section>
            <div
              class="MuiBox-root css-dvxtzn"
            >
              <button
                class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root css-124vkh3-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
                data-testid="login-submit"
                id="mui-1"
                tabindex="0"
                type="submit"
              >
                Login
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButtonBase-root  css-1d9tfpj-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Or Sign up for an account
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </form>
          <div
            class="MuiBox-root css-h8m1q6"
          >
            <a
              class="MuiTypography-root MuiTypography-subtitle1 MuiLink-root MuiLink-underlineAlways css-1vmwua3-MuiTypography-root-MuiLink-root"
              data-testid="forgot-password-link"
              href="/forgot-password"
            >
              Forgot password
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-1nrbtl5"
    >
      <div
        class="MuiBox-root css-mf869q"
      >
        <div
          height="80"
          width="80"
        />
      </div>
      <div
        class="MuiBox-root css-1cb30mv"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-1ujikh7-MuiTypography-root"
        >
          Log in to Team Teach
        </h6>
        <form
          autocomplete="off"
          class="MuiBox-root css-h5fkc8"
          data-testid="LoginForm"
          novalidate=""
        >
          <section
            class="MuiBox-root css-0"
          >
            <div
              class="MuiBox-root css-1yjvs5a"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                title="Email Address"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="email"
                  id="email-label"
                >
                  Email Address
                </label>
                <div
                  class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-ghsjzk-MuiInputBase-root-MuiInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                    data-testid="input-email"
                    id="email"
                    name="email"
                    placeholder="Please enter your email address"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiBox-root css-1yjvs5a"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                title="Password"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="password"
                  id="password-label"
                >
                  Password
                </label>
                <div
                  class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-ghsjzk-MuiInputBase-root-MuiInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                    data-testid="input-password"
                    id="password"
                    name="password"
                    placeholder="Please enter your password"
                    type="password"
                    value=""
                  />
                </div>
              </div>
            </div>
          </section>
          <div
            class="MuiBox-root css-dvxtzn"
          >
            <button
              class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root css-124vkh3-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
              data-testid="login-submit"
              id="mui-1"
              tabindex="0"
              type="submit"
            >
              Login
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButtonBase-root  css-1d9tfpj-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Or Sign up for an account
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
        <div
          class="MuiBox-root css-h8m1q6"
        >
          <a
            class="MuiTypography-root MuiTypography-subtitle1 MuiLink-root MuiLink-underlineAlways css-1vmwua3-MuiTypography-root-MuiLink-root"
            data-testid="forgot-password-link"
            href="/forgot-password"
          >
            Forgot password
          </a>
        </div>
      </div>
    </div>
  </div>,
  "context": Object {
    "auth": Object {
      "loading": false,
      "login": [MockFunction],
      "logout": [MockFunction],
      "profile": Object {
        "addresses": Array [],
        "attributes": Array [],
        "contactDetails": Array [
          Object {
            "email": "john.smith@example.com",
          },
        ],
        "createdAt": "",
        "familyName": "Smith",
        "givenName": "John",
        "id": "cacb559d-b85d-5e64-b623-37252520ebda",
        "preferences": Array [],
        "status": "active",
        "tags": Array [],
        "title": "Mr",
        "updatedAt": "",
      },
    },
  },
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
